### PR TITLE
RavenDB-19985 Switch to FluentFTP

### PR DESCRIPTION
**Pull request by itayweb**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19985/Switch-to-FluentFTP

### Additional description

First test current functionality by creating 2 slow tests and then after those passed, migrate the raven ftp client to use FluentFTP library and tested using those 2 slow tests with some minor changes to use FluentFTP commands

### Type of change
- Task

### How risky is the change?
- Low 

### Is it platform specific issue?

- Yes, v6.0 and above

### Documentation update
- No documentation update is needed 

### Testing by Contributor
- Tests have been added that prove the fix is effective or that the feature works

### UI work
- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 1]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
+                    try
                     {
-                        // folder already exists
-                        continue;
+                        if (!client.DirectoryExists($"/{directory}"))
+                            client.CreateDirectory($"/{directory}");
+                    }
+                    catch (Exception e)
+                    {
+                        throw new Exception();
```
Comment (garayx) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: lets throw the original exception here

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 2]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
+                    client.Connect();
+                }
+                catch (AuthenticationException e)
+                {
+                    throw new AuthenticationException("Make sure that you provided the correct certificate " +
+                                                      "and that the url matches the one that is defined in it", e);
+                }
+                catch (Exception e)
+                {
+                    throw new Exception();
```
Comment (garayx) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: lets throw the original exception here

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        public async Task CanUploadBackup()
+        {
+            var services = new ServiceCollection();
+            services.Configure<InMemoryFileSystemOptions>(opt => opt.KeepAnonymousFileSystem = true);
+            services.AddFtpServer(builder => builder.UseInMemoryFileSystem().EnableAnonymousAuthentication());
+            services.Configure<FtpServerOptions>(opt => opt.ServerAddress = "127.0.0.1");
+            await using (var serviceProvider = services.BuildServiceProvider())
+            {
+                var ftpServerHost = serviceProvider.GetRequiredService<IFtpServerHost>();
+                ftpServerHost.StartAsync(CancellationToken.None).Wait();
```
Comment (garayx) [inline test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: please use await instead of `Wait()` on all async methods

**Review by garayx — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 1 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 2 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs (hunk 3 above)
Comment (itayweb) [reply test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 4]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
             var url = CreateNestedFoldersIfNeeded(folderName);
             url += $"/{fileName}";
 
-            var request = CreateFtpWebRequest(url, WebRequestMethods.Ftp.UploadFile, keepAlive: true);
-            var readBuffer = new byte[DefaultBufferSize];
-
-            int count;
-            var requestStream = request.GetRequestStream();
-            while ((count = stream.Read(readBuffer, 0, readBuffer.Length)) != 0)
+            using (var client = CreateFtpClient(_url, keepAlive: true))
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: we should add error handling to this method, check other clients, how they handle uploads, how the handle progress state change

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 5]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
         }
 
         private string CreateNestedFoldersIfNeeded(string folderName)
         {
-            ExtractUrlAndDirectories(out var url, out var directories);
+            ExtractUrlAndDirectories(out var directories);
 
             // create the nested folders including the new folder
             directories.Add(folderName);
+            string url = "";
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: string.Empty

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 6]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
-                    var response = (FtpWebResponse)e.Response;
-                    if (response.StatusCode == FtpStatusCode.ActionNotTakenFileUnavailable)
+                    try
                     {
-                        // folder already exists
-                        continue;
+                        if (!client.DirectoryExists($"/{directory}"))
+                            client.CreateDirectory($"/{directory}");
+                    }
+                    catch (WebException e)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: `e` is not used,
you are catching an exception and not doing anything with it, so why bother?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 7]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
+            client.Credentials = new NetworkCredential(_userName, _password);
             if (_useSsl)
             {
                 var byteArray = Convert.FromBase64String(_certificateAsBase64);
 
                 try
                 {
-                    var x509Certificate = new X509Certificate(byteArray);
-                    request.ClientCertificates = new X509CertificateCollection(new[] { x509Certificate });
+                    var x509Certificate = new X509Certificate2(byteArray);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:176]: it would be great to have a test for this case

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 8]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
-            request.UseBinary = true;
-            request.KeepAlive = keepAlive;
 
-            return request;
+            client.Config.SocketKeepAlive = keepAlive;
+            if (_port != null)
+                client.Port = (int)_port;
+            else
+                client.Port = DefaultFtpPort;
+            client.Connect();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:195]: IRIC there is a method AutoConnect or sth like that, maybe it is better to use it? What it does?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 9]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
-            }
-            catch (WebException e)
-            {
-                var response = (FtpWebResponse)e.Response;
-                if (response.StatusCode == FtpStatusCode.ActionNotTakenFileUnavailable)
+                try
                 {
-                    // folder doesn't exist
-                    return;
+                    client.Connect();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:225]: maybe auto connect?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 10]
```diff
@@ -65,135 +67,121 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
+                }
+                catch (AuthenticationException e)
+                {
+                    throw new AuthenticationException("Make sure that you provided the correct certificate " +
+                                                      "and that the url matches the one that is defined in it", e);
+                }
+                catch (WebException e)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: you are not doing anything with this exception

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+                    Password = "itay@ravendb.net"
+                };
+                using (var client = new RavenFtpClient(settings))
+                using (var stream = new MemoryStream(Encoding.UTF8.GetBytes("abc")))
+                {
+                    client.UploadFile("testFolder", "testFile", stream);
+                    var isExist = CheckFile(settings.Url, "testFolder", "testFile", client);
+                    Assert.Equal(true, isExist);
+                }
+                await ftpServerHost.StopAsync(CancellationToken.None);
```
Comment (ppekrol) [inline test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: can this be in some kind of a `try/finally` block?

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+                    var backupResult = (BackupResult)store.Maintenance.Send(new GetOperationStateOperation(await Backup.GetBackupOperationIdAsync(store, backupId))).Result;
+                    var lastEtag = store.Maintenance.Send(new GetStatisticsOperation()).LastDocEtag;
+                    var status = await Backup.RunBackupAndReturnStatusAsync(Server, backupId, store,  expectedEtag: lastEtag);
+                    var isExist = CheckBackupFile(settings.Url, status.FolderName, client);
+                    Assert.NotNull(backupResult);
+                    Assert.Equal(UploadState.Done, backupResult.FtpBackup.UploadProgress.UploadState);
+                    Assert.Equal(true, isExist);
+                }
+
+                await ftpServerHost.StopAsync(CancellationToken.None);
```
Comment (ppekrol) [inline test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: `try/finally` ?

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+using Raven.Client.Documents.Operations.Backups;
+using Raven.Server.Documents.PeriodicBackup;
+using Raven.Tests.Core.Utils.Entities;
+using Tests.Infrastructure;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace SlowTests.Server.Documents.PeriodicBackup
+{
+    public class Ftp : CloudBackupTestBase
```
Comment (ppekrol) [inline test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: assuming that those tests with the old client were passing as well?

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+                    client.UploadFile("testFolder", "testFile", stream);
+                    var isExist = CheckFile(settings.Url, "testFolder", "testFile", client);
+                    Assert.Equal(true, isExist);
+                }
+                await ftpServerHost.StopAsync(CancellationToken.None);
+            }
+        }
+
+        [FtpFact]
+        public async Task CanUploadBackup()
```
Comment (ppekrol) [inline test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: any other tests to migrate?

**Review by ppekrol — COMMENTED** [review]
Looks great. Maybe we want to support other features now that were not supported by FTP? Like retention policies, etc. What other clients support?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 5 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs (hunk 11 above)
Comment (itayweb) [reply test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: Done

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs (hunk 12 above)
Comment (itayweb) [reply test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: Done

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs (hunk 13 above)
Comment (itayweb) [reply test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: Correct, with a bit of change because of the difference between the 2 libraries (FluentFTP and FTPWebRequest)

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 6 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 10 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 8 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:195]: It's automatically searching for FTP settings and if it finds one it automatically connecting otherwise returning a null, I will convert to AutoConnect method

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 9 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:225]: Done

Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:225]: AutoConnect caused problems with the tests were Connect didn't

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 8 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:195]: AutoConnect caused problems with the tests were Connect didn't

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 7 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:176]: Done

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs (hunk 14 above)
Comment (itayweb) [reply test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 7 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs:176]: Done

File: test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs [hunk 15]
```diff
@@ -0,0 +1,375 @@
+﻿using System;
```
Comment (garayx) [inline test/SlowTests/Server/Documents/PeriodicBackup/Ftp.cs]: rename the file  to `FtpBasicTests`

File: test/Tryouts/Program.cs [hunk 16]
```diff
@@ -8,6 +8,7 @@
 using Raven.Server.Utils;
 using SlowTests.Cluster;
 using SlowTests.Issues;
+using SlowTests.Server.Documents.PeriodicBackup;
```
Comment (garayx) [inline test/Tryouts/Program.cs:11]: revert the changes in this file

**Review by garayx — APPROVED** [review]

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 17]
```diff
@@ -130,72 +139,110 @@ private void ExtractUrlAndDirectories(out string url, out List<string> dirs)
...
 
                 try
                 {
-                    var x509Certificate = new X509Certificate(byteArray);
-                    request.ClientCertificates = new X509CertificateCollection(new[] { x509Certificate });
+                    var x509Certificate = new X509Certificate2(byteArray);
+                    client.Config.EncryptionMode = FtpEncryptionMode.Explicit;
+                    client.Config.SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls13;
+                    client.Config.ClientCertificates.Add(x509Certificate);
+                    client.Config.ValidateAnyCertificate = true;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: I do not think this should be set to `true` - it means that certificate validation will not be performed, right?
Probably you want to set it in test to `true` because our self-signed certificate validation fails, but in real life this should not be set.

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 18]
```diff
@@ -130,72 +139,110 @@ private void ExtractUrlAndDirectories(out string url, out List<string> dirs)
...
             {
                 var byteArray = Convert.FromBase64String(_certificateAsBase64);
 
                 try
                 {
-                    var x509Certificate = new X509Certificate(byteArray);
-                    request.ClientCertificates = new X509CertificateCollection(new[] { x509Certificate });
+                    var x509Certificate = new X509Certificate2(byteArray);
+                    client.Config.EncryptionMode = FtpEncryptionMode.Explicit;
+                    client.Config.SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls13;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: do we need to set it?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 19]
```diff
@@ -130,72 +139,110 @@ private void ExtractUrlAndDirectories(out string url, out List<string> dirs)
...
+        }
+
+        public void DeleteFolder(string folderName)
+        {
+            ExtractUrlAndDirectories(out var url, out _);
+            using (var client = CreateFtpClient(url, keepAlive: false))
             {
-                throw new AuthenticationException("Make sure that you provided the correct certificate " +
-                                                  "and that the url matches the one that is defined in it", e);
+                client.DeleteDirectory($"{folderName}");
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: why interpolating this string?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 20]
```diff
@@ -130,72 +139,110 @@ private void ExtractUrlAndDirectories(out string url, out List<string> dirs)
...
+            return GetItemsInternal(url, path, FtpObjectType.Directory);
+        }
+
+        public List<string> GetFiles(string folderName)
+        {
+            ExtractUrlAndDirectories(out var url, out _);
+            return GetItemsInternal(url, folderName, FtpObjectType.File);
+        }
+
+        public void DeleteFolder(string folderName)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: always good to assert argument values, VS or Resharper will be helpful, you should assert at least for null, but probably for null or whitespace?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 21]
```diff
@@ -130,72 +139,110 @@ private void ExtractUrlAndDirectories(out string url, out List<string> dirs)
...
+
+        public List<string> GetFolders()
+        {
+            TestConnection();
+            ExtractUrlAndDirectories(out var url, out _);
+            var path = GetPath();
+            return GetItemsInternal(url, path, FtpObjectType.Directory);
+        }
+
+        public List<string> GetFiles(string folderName)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: assert folderName (check other comment)

File: src/Raven.Server/Documents/PeriodicBackup/Retention/FtpRetentionPolicyRunner.cs [hunk 22]
```diff
@@ -1,13 +1,15 @@
...
+using Raven.Server.Documents.Indexes.Static.Extensions;
 
 namespace Raven.Server.Documents.PeriodicBackup.Retention
 {
     public class FtpRetentionPolicyRunner : RetentionPolicyRunnerBase
     {
         private readonly RavenFtpClient _client;
 
-        protected override string Name => "Glacier";
+        protected override string Name => "Ftp";
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/Retention/FtpRetentionPolicyRunner.cs]: FTP

File: test/SlowTests/Server/Documents/PeriodicBackup/FtpBasicTests.cs [hunk 23]
```diff
@@ -0,0 +1,375 @@
...
+                    BackupConfigurationHelper.SkipMinimumBackupAgeToKeepValidation = false;
+                    await ftpServerHost.StopAsync(CancellationToken.None);
+                }
+            }
+        }
+
+        [FtpFact]
+        public async Task CanUploadFileOnEncrypted()
+        {
+            CertificateUtils.CreateCertificateAuthorityCertificate("auth", out var caKey, out var caName);
```
Comment (ppekrol) [inline test/SlowTests/Server/Documents/PeriodicBackup/FtpBasicTests.cs]: Wouldn't it be better to use Certificates.GenerateAndSaveSelfSignedCertificate method? this way you are using cached one instead of generating it from scratch each time (which takes time)

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 19 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 20 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 21 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/Retention/FtpRetentionPolicyRunner.cs (hunk 22 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/Retention/FtpRetentionPolicyRunner.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 17 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 18 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Removed

File: test/SlowTests/Server/Documents/PeriodicBackup/FtpBasicTests.cs (hunk 23 above)
Comment (itayweb) [reply test/SlowTests/Server/Documents/PeriodicBackup/FtpBasicTests.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 24]
```diff
@@ -62,63 +70,68 @@ public void UploadFile(string folderName, string fileName, Stream stream)
...
+                    var readBuffer = new byte[DefaultBufferSize];
 
-            Progress?.UploadProgress.ChangeState(UploadState.PendingResponse);
-            using (request.GetResponse())
-            {
-                Progress?.UploadProgress.ChangeState(UploadState.Done);
+                    int count;
+                    while ((count = stream.Read(readBuffer, 0, readBuffer.Length)) != 0)
+                    {
+                        client.UploadBytes(readBuffer, path, FtpRemoteExists.Resume, progress: (p) =>
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: how does the progress behave? you have a while loop here and start to upload another chunk, wouldn't it set the value to 0 again? can't it be done via 1 call?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 25]
```diff
@@ -130,72 +143,126 @@ private void ExtractUrlAndDirectories(out string url, out List<string> dirs)
...
         {
-#pragma warning disable SYSLIB0014
-            var request = (FtpWebRequest)WebRequest.Create(new Uri(url));
-#pragma warning restore SYSLIB0014
-            request.Method = method;
-
-            request.Credentials = new NetworkCredential(_userName, _password);
-            request.EnableSsl = _useSsl;
+            var client = new FtpClient(url);
+            client.Config.ConnectTimeout = -1;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: this can hang now forever, what better value we can put here?

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 26]
```diff
@@ -53,6 +60,7 @@ public RavenFtpClient(FtpSettings ftpSettings, Progress progress = null, Cancell
                 _url += "/";
 
             Debug.Assert(_url.StartsWith("ftp://", StringComparison.OrdinalIgnoreCase));
+            _isTesting = Environment.GetEnvironmentVariable("isTesting") != null;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: not a good idea, would prefer a local field or a static set in the TestBase or RavenTestBase (in static ctor)

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs [hunk 27]
```diff
@@ -130,72 +143,126 @@ private void ExtractUrlAndDirectories(out string url, out List<string> dirs)
...
+
+        private static void OnValidateCertificate(BaseFtpClient control, FtpSslValidationEventArgs e)
+        {
+            if (e.PolicyErrors == SslPolicyErrors.None)
+            {
+                e.Accept = true;
+            }
+            else
+            {
+                throw new Exception($"{e.PolicyErrors}{Environment.NewLine}{e.Certificate}");
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: IIRC, there is a AuthenticationException available, wouldn't it be better? Also some message would be nice for the user so it would be easier to nail down the problem?

Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 25 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 26 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

File: src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs (hunk 24 above)
Comment (itayweb) [reply src/Raven.Server/Documents/PeriodicBackup/RavenFtpClient.cs]: Done

**Review by ppekrol — APPROVED** [review]
